### PR TITLE
cli/sql: report the cluster ID and server version upon new connections

### LINKED ISSUE #14310: cli: show node that sql client is connected to
**Feature request**

 When you start the built-in sql shell, you get a welcome message with some simple tips:

```
root@load-balance-test-ubuntu-2gb-nyc3-02:~# cockroach sql --host=138.197.78.223 --port=26257 --insecure
# Welcome to the cockroach SQL interface.
# All statements must be terminated by a semicolon.
# To exit: CTRL + D.
```

It'd be nice to add details about the node connected to. This would be especially useful when you are using a load balancer and want to validate that the connections are getting proxied to different nodes.

@arjunravinarayan

### PR DESCRIPTION
**Pull request by knz**
### sql: expose the node ID of the gateway node as a read-only session variable

The new variable is called `node_id`.

This makes it more convenient than `SELECT node_id FROM
crdb_internal.node_build_info LIMIT 1`.

Requested/suggested by @jseldess.

Fixes #14310.

### cli/sql: report the cluster ID and server version upon new connections

This patch ensures the client prints out the cluster ID and
server version, and if different from the server version, also the
client version, both upon the initial connection and upon reconnects
*if they change*.

For example:

    $ ./cockroach sql
    # Welcome to the cockroach SQL interface.
    # All statements must be terminated by a semicolon.
    # To exit: CTRL + D.
    #
    # Server version: CockroachDB CCL 0da9d8f-dirty (linux amd64, built 2017/06/23 19:50:15, go1.8) (same version as client)
    # Cluster ID: fe26b6f3-3153-4c75-a7fb-3027c327e5ab
    #
    # Enter \? for a brief introduction.
    #
    root@localhost:26257/>

Then upon a reconnect, if the node ID changes:

    root@localhost:26257/> SELECT 1;
    driver: bad connection
    connection lost; opening new connection: all session settings will be lost
    root@localhost:26257/>

If the server version has changed:

    root@localhost:26257/> select 1;
    driver: bad connection
    connection lost; opening new connection: all session settings will be lost
    # Client version: CockroachDB CCL a1b6734-dirty (linux amd64, built 2017/06/23 19:14:15, go1.8)
    # Server version: CockroachDB CCL a1b6734-dirty (linux amd64, built 2017/06/23 19:15:00, go1.8)
    root@localhost:26257/>

If the cluster ID has changed:

        root@localhost:26257/> SELECT 1;
        driver: bad connection
        connection lost; opening new connection: all session settings will be lost
        error retrieving the transaction status: the cluster ID has changed!
        Previous ID: b8d0cc65-3a71-4db8-bcda-6fa86206ebb0
        New ID: fe26b6f3-3153-4c75-a7fb-3027c327e5ab
        root@localhost:26257/ ?>

Supersedes #16615.

cc @mberhault

### CODE REVIEW

**Comment by knz** [comment]
Marc I'm assinging to you because you had an interest in the future and Matt is out this week. Feel free to de-assign if this feels out of your league.

File: pkg/cli/sql_util.go [hunk 1]
```diff
@@ -71,10 +80,143 @@ func (c *sqlConn) ensureConn() error {
...
+	}
+
+	// Report the cluster ID only if it it could be fetched
+	// successfully, and it has changed since the last connection.
+	if newClusterID != c.clusterID && newClusterID != "" {
+		// As a special case, we complain loudly if the cluster ID
+		// actually changes. This is sign of a potentially very bad load
+		// balancer configuration.
+		if c.clusterID != "" {
+			fmt.Print("\n# WARNING: the cluster ID has changed!\n\n")
```
Comment (mberhault) [inline pkg/cli/sql_util.go]: This isn't enough, we should error out if reconnecting to a different cluster. I'd even be happy with a hard `log.Fatal`

File: pkg/cli/sql_util.go [hunk 2]
```diff
@@ -71,10 +80,143 @@ func (c *sqlConn) ensureConn() error {
...
+		}
+		c.clusterID = newClusterID
+		fmt.Println("# Cluster ID:", newClusterID)
+	}
+
+	// Report the node ID only if it it could be fetched successfully,
+	// and it has changed since the last connection.
+	if newNodeID != c.nodeID && newNodeID != "" {
+		c.nodeID = newNodeID
+		fmt.Println("# Node ID:", newNodeID)
```
Comment (mberhault) [inline pkg/cli/sql_util.go]: I still don't think this is necessary. We should document the contents of `crdb_internal` for people who care, but we definitely don't need to log this for most users.

File: pkg/cli/sql_util.go [hunk 3]
```diff
@@ -71,10 +80,143 @@ func (c *sqlConn) ensureConn() error {
...
+
+	// Report the server version only if it the revision has been
+	// fetched successfully, and the revision has changed since the last
+	// connection.
+	if newServerBuildInfo != c.serverBuildInfo && newServerBuildInfo.Revision != "?" {
+		c.serverBuildInfo = newServerBuildInfo
+
+		serverVersionShort := newServerBuildInfo.Short()
+		isSame := ""
+		if serverVersionShort == clientVersionShort {
```
Comment (mberhault) [inline pkg/cli/sql_util.go]: I would limit the "sameness" check to the tag. We don't care if they're different platforms/build tools/timestamps. The important bit is the wire protocol.

**Comment by knz** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cli/sql_util.go, line 159 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-159:-KnZTmveaZKch9QZhDZR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L159)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I would limit the "sameness" check to the tag. We don't care if they're different platforms/build tools/timestamps. The important bit is the wire protocol.
</blockquote></details>

Done.

---

*[pkg/cli/sql_util.go, line 176 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-176:-KnZTnV93Ft7wYnsjuXv:b-8k12a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L176)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

This isn't enough, we should error out if reconnecting to a different cluster. I'd even be happy with a hard `log.Fatal`
</blockquote></details>

`log.Fatal` is not suitable: it's common (at least in my workflow, but I know I'm not the only one) to leave an interactive SQL shell open across test cluster resets. 

Instead I changed to use an error, which prevents the query from being executed.

---

*[pkg/cli/sql_util.go, line 187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-187:-KnZVNEP1hnTcFQSlLoq:b-u5o6sl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L187)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I still don't think this is necessary. We should document the contents of `crdb_internal` for people who care, but we definitely don't need to log this for most users.
</blockquote></details>

That's fine by me. @jseldess can you confirm that #14310 would be suitably addressed by recommending a SELECT query into `crdb_internal.node_build_info` instead of having the CLI shell report the node ID in interactive sessions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703)*

**Comment by jseldess** [comment]
*[pkg/cli/sql_util.go, line 187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-187:-KnZpmJV5Z5nuhLihdYT:bkg0zmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L187)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That's fine by me. @jseldess can you confirm that #14310 would be suitably addressed by recommending a SELECT query into `crdb_internal.node_build_info` instead of having the CLI shell report the node ID in interactive sessions?
</blockquote></details>

Sure, I'm also fine leaving it out. We already do recommend querying that internal table in a few places, for example: https://www.cockroachlabs.com/docs/manual-deployment.html#step-6-test-load-balancing. In the referenced issue, we also discussed a session variable for getting the node ID...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703)*

File: pkg/cli/sql_util.go (hunk 2 above)
Comment (knz) [inline pkg/cli/sql_util.go]: ok, removed. Added a session variable as recommended/suggested by Jesse.

File: pkg/cli/sql_util.go [hunk 4]
```diff
@@ -69,12 +79,160 @@ func (c *sqlConn) ensureConn() error {
...
+	}
+	if err != nil {
+		fmt.Fprintln(stderr, "unable to retrieve the server's version")
+	} else {
+		defer func() { _ = rows.Close() }()
+
+		// Read the node_build_info table as an array of strings.
+		rowVals, err := getAllRowStrings(rows, true /* showMoreChars */)
+		if err != nil || len(rowVals) == 0 || len(rowVals[0]) != 3 {
+			fmt.Fprintln(stderr, "error while rerieving the server's version")
```
Comment (mberhault) [inline pkg/cli/sql_util.go]: s/rerieving/retrieving/

**Review by mberhault — APPROVED** [review]
LGTM, though you haven't pushed your changes to the last round of comments.

File: pkg/cli/sql_util.go (hunk 4 above)
Comment (knz) [inline pkg/cli/sql_util.go]: Done.

**Comment by knz** [comment]
I did push the changes! They are updated at the source branch https://github.com/knz/cockroach/tree/20170623-cli-ver-reporting

somehow GH is not picking it up here?!

**Comment by knz** [comment]
ok it's picked up now :)

**Comment by knz** [comment]
ok that went through! TFYRs

**Pull request merged**